umh-core/pkg/fsm/nmap/reconcile.go

```diff
@@ -0,0 +1,444 @@
+		}
+		// Send event to transition from Stopped to Starting
+		return n.baseFSMInstance.SendEvent(ctx, EventStart), true
+	case IsStartingState(currentState):
+		return n.reconcileStartingState(ctx, filesystemService, currentState, currentTime)
+	case IsRunningState(currentState):
+		return n.reconcileRunningState(ctx, filesystemService, currentState, currentTime)
+	}
+
+	return nil, false
+}
+
+// reconcileTransitionToStopped handles transitions when the desired state is Stopped.
+// Transition sequence for running state:
+//
+//	from any running state (open, filtered, closed, degraded) -> stopping (EventStop)
+//	then from stopping -> stopped (EventStopped)
+func (n *NmapInstance) reconcileTransitionToStopped(ctx context.Context, currentState string, filesystemService filesystem.Service) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentNmapMonitor, n.baseFSMInstance.GetID()+".reconcileTransitionToStopped", time.Since(start))
+	}()
+
+	// If not yet stopped and not already stopping, initiate stop.
+	if currentState != OperationalStateStopped && currentState != OperationalStateStopping {
+		if err := n.monitoringStop(ctx); err != nil {
+			return err, true
+		}
+		return n.baseFSMInstance.SendEvent(ctx, EventStop), true
+	}
+
+	// If currently stopping, verify that it is now fully stopped.
+	if currentState == OperationalStateStopping {
+		if n.IsStopped() {
+			return n.baseFSMInstance.SendEvent(ctx, EventStopDone), true
+		}
+		return nil, false
+	}
+
+	return nil, false
+}
+
+// reconcileStartingState handles the various starting phase states when transitioning to Active.
+func (n *NmapInstance) reconcileStartingState(ctx context.Context, filesystemService filesystem.Service, currentState string, currentTime time.Time) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentBenthosInstance, n.baseFSMInstance.GetID()+".reconcileStartingState", time.Since(start))
+	}()
+
+	switch currentState {
+	case OperationalStateStarting:
+		return n.baseFSMInstance.SendEvent(ctx, EventStartDone), true
+	default:
+		return fmt.Errorf("invalid starting state: %s", currentState), false
+	}
+}
+
+// reconcileRunningState handles the various running states when transitioning to Active.
+func (n *NmapInstance) reconcileRunningState(ctx context.Context, filesystemService filesystem.Service, currentState string, currentTime time.Time) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentBenthosInstance, n.baseFSMInstance.GetID()+".reconcileRunningState", time.Since(start))
+	}()
+
+	switch currentState {
+	case OperationalStateOpen:
+		// If we're in open, we need to check whether it is degraded
+		if !n.isNmapHealthy() {
+			return n.baseFSMInstance.SendEvent(ctx, EventNmapExecutionFailed), true
+		}
+
+		changedPortState, event := n.checkPortState(currentState)
```

Comment:
add comment:

// if not degraded, check if the last reported state was open/filtered/closed and then set its state accordingly